File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/SeatingPlace.java
package org.weaveit.seatingplacesuggestions;

public record SeatingPlace(
        String rowName,
        int number,
        PricingCategory pricingCategory,
        SeatingPlaceAvailability seatingPlaceAvailability
) {
    public boolean isAvailable() {
        return seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE;
    }

    public boolean matchCategory(PricingCategory pricingCategory) {
        if (pricingCategory == PricingCategory.MIXED) {
            return true;
        }
        return this.pricingCategory == pricingCategory;
    }

    public SeatingPlace allocate() {
        if (seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE) {
            return new SeatingPlace(rowName, number, pricingCategory, SeatingPlaceAvailability.ALLOCATED);
        }
        return this;
    }

    public String name() {
        return rowName + number;
    }

    @Override
    public String toString() {
        return name();
    }
}

File: TheaterSuggestions/Kotlin/ScreeningSeatingArrangementSuggestions/src/main/kotlin/org/weaveit/seatingplacesuggestions/SeatingPlace.kt
package org.weaveit.seatingplacesuggestions

class SeatingPlace(
    private val rowName: String,
    private val number: Int,
    private val pricingCategory: PricingCategory,
    private val isAvailable: Boolean
) {
    fun isAvailable(): Boolean = isAvailable

    fun matchCategory(pricingCategory: PricingCategory): Boolean = this.pricingCategory == pricingCategory

    fun name(): String = "$rowName$number"
}
